### ISSUE TITLE
Broken async/await in the VM

### ISSUE THREAD

**Issue by natebosch** [dart-lang/sdk#33660]
It's pretty tough to narrow in on a minimal repro, here's the simplest I've been able to get:

```dart
import 'dart:async';

void main() async {
  var t = new Node('outer', [
    new Node('missing'),
    new Node('a', [
      new Node('b', [
        new Node('c'),
        new Node('d', [new Node('e')]),
        await new Future.value(new Node('async_inner')),
      ]),
    ]),
    await new Future.value(new Node('async_outer')),
  ]);
  print(t);
}

class Node {
  final List<Node> nested;
  final String name;

  Node(this.name, [this.nested]);

  @override
  String toString() => '<$name:[${nested?.join('\n')}]>';
}
```

Expected output:

```
<outer:[<missing:[null]>
<a:[<b:[<c:[null]>
<d:[<e:[null]>]>
<async_inner:[null]>]>]>
<async_outer:[null]>]>
```

VM output:

```
<outer:[<c:[null]>
<a:[<b:[<c:[null]>
<d:[<e:[null]>]>
<async_inner:[null]>]>]>
<async_outer:[null]>]>
```

Note that the first `missing` has been replaced by an extra `c`.

The VM produces the expected output with `--no-preview-dart-2`. Dart2js (at least in dartpad, should be using the CFE) [produces the correct output](https://dartpad.dartlang.org/25f8061884825ac19b9807bf1f551a4b)

**Comment by natebosch** [comment]
cc @a-siva @kmillikin

**Comment by jakemac53** [comment]
I checked this with dartdevc and dartdevk and both work fine, which indicates to me this is probably a vm specific issue.

**Comment by natebosch** [comment]
Taking a look at a script snapshot from the VM we can see it reusing the variable `:async_temporary_1` for both the `'missing'` node and `'c'` node which is causing the problem. Since other backends don't exhibit this behavior this may be something do with the way the VM is using the frontend, but it does look like a kernel issue.

**Comment by vsmenon** [comment]
@a-siva - the "missing" is showing with ddk and dart2js.  @natebosch 's comment above in addition seems to point at a backend problem with the vm.

**Comment by natebosch** [comment]
I can see [the bug in the kernel](https://gist.github.com/natebosch/5972ffb1a80fad9577417611cd45aa91#file-broken_async-dill-L29-L30) when I run `dart --snapshot_kind=script --snapshot=out.dill` and then dump the kernel.

**Comment by jakemac53** [comment]
> I can see the bug in the kernel when I run dart --snapshot_kind=script --snapshot=out.dill and then dump the kernel.

Sounds like a kernel transformation that the vm is applying then?

**Comment by kmillikin** [comment]
There is a Kernel to Kernel transformation that tries to minimize the number of temporary variables it uses to name intermediate expression.  It sounds like it has a bug.  I'll take a look tomorrow.

**Comment by JekCharlsonYu** [comment]
ping!

**Comment by jensjoha** [comment]
@kmillikin has been OOO.
I'm adding tests in https://dart-review.googlesource.com/c/sdk/+/63821 and have a crude fix in https://dart-review.googlesource.com/c/sdk/+/63823. The fix is not great though and we should probably find a better solution asap.
I'm unfamiliar with the code, but will give it a try...

**Comment by jensjoha** [comment]
Fixed with 3a1b09dc63542384e2c4a6016a0e3486fda4c4d4.
